### PR TITLE
Merge tarfile

### PR DESCRIPTION
**Pull request by tharvik**
https://docs.python.org/3/library/tarfile.html

Failling on pytype because of [pytype#35](https://github.com/google/pytype/issues/35).

Also add `ContextManager` to `typing`.

### CODE REVIEW

**Comment by tharvik** [comment]
There is also an issue with mypy and I don't know why (maybe an issue with importing loop between `typing` and `types`)
```
stdlib/2.7/types.pyi:16: error: Cannot determine type of 'long'
```

**Comment by gvanrossum** [comment]
Just replace the long with int, long is just an alias for int anyways in mypy.

**Comment by tharvik** [comment]
Using `long` works, there is still the pytype bug though.

**Comment by gvanrossum** [comment]
Hey @tharvik and @matthiaskramm -- if you added this file to the pytype blacklist, could it be merged?

**Comment by matthiaskramm** [comment]
Yeah, pytype needs `elif` support. Please add `tarfile.pyi` to `tests/pytype_blacklist.txt` for the time being.

**Comment by tharvik** [comment]
Done in 6c8e796.

File: stdlib/2.7/typing.pyi [hunk 1]
```diff
@@ -174,6 +175,15 @@ class ValuesView(MappingView, Iterable[_VT_co], Generic[_VT_co]):
     def __contains__(self, o: object) -> bool: ...
     def __iter__(self) -> Iterator[_VT_co]: ...
 
+class ContextManager(Generic[_T]):
```
Comment (gvanrossum) [inline stdlib/2.7/typing.pyi]: Why did you add ContextManager to the 2.7/typing.pyi? It's not in the python2.7/typing.py in the [python/typing](https://github.com/python/typing) repo. And even for Python 3 it's only present from 3.6 on (because it derives from contextlib.AbstractContextManager which is new in 3.6).

File: stdlib/2and3/tarfile.pyi [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        *, format: Optional[int] = ..., tarinfo: Optional[TarInfo] = ...,
+        dereference: Optional[bool] = ...,
+        ignore_zeros: Optional[bool] = ...,
+        encoding: Optional[str] = ..., errors: str = ...,
+        pax_headers: Optional[Mapping[str, str]] = ...,
+        debug: Optional[int] = ...,
+        errorlevel: Optional[int] = ...) -> TarFile: ...
+
+
+class TarFile(ContextManager[TarFile], Iterable):
```
Comment (gvanrossum) [inline stdlib/2and3/tarfile.pyi]: FWIW there's no strong need to inherit from ContextManager -- mypy accepts `with` statements using duck typing, i.e. it just looks whether the object defines `__enter__` and `__exit__`.

File: stdlib/2.7/typing.pyi (hunk 1 above)
Comment (tharvik) [inline stdlib/2.7/typing.pyi]: It is in the PEP 484, so I tough it was for any version; why having it only for 3.6 anyway?
In this case, it way to allow having `tarfile` in `2and3`.

File: stdlib/2and3/tarfile.pyi (hunk 2 above)
Comment (tharvik) [inline stdlib/2and3/tarfile.pyi]: True; but that's a mypy-specific extension, IMO, that's not explicitly in the PEP that it should be supported. I can remove it if prefered.

**Comment by gvanrossum** [comment]
Can you first send a PR to the https://github.com/python/typing repo to add typing.ContextManager there for versions before 3.6? The easiest way to do that would probably to define AbstractContextManager locally (as a copy of the one in contextlib in 3.6) if sys.version_info < 3.6.

File: stdlib/2and3/tarfile.pyi (hunk 2 above)
Comment (gvanrossum) [inline stdlib/2and3/tarfile.pyi]: I would prefer removing it for now. The same thing is done all over typeshed -- essentially ContextManager is recognized by `with` via duck typing, and `for` recognizes iterables the same way.

Comment (tharvik) [inline stdlib/2and3/tarfile.pyi]: Removed.

**Comment by tharvik** [comment]
Then, what's the use of `ContextManager` if is should not be used in the stub? Enforce an implementation of `__enter__` and `__exit__`? Then what's the use of `AbstractContextManager`?

**Comment by gvanrossum** [comment]
There never was a ContextManager ABC in the stdlib, which is why its introduction has a few hiccups. You can mostly do without it (just define `__enter__` and `__exit__` and use it in a `with` statement) but sometimes you want to say "this returns a context manager". Eventually, in 3.6, contextlib.AbstractContextManager was added to the stdlib. We're now figuring out how to deal with this -- for now, we only have typing.ContextManager inheriting from contextlib.AbstractContextManager in 3.6, but we should just have it for all versions, and remove contextlib.ContextManager from contextlib.pyi in typeshed (because it doesn't correspond to anything in the stdlib).

I hope that's clearer -- if not please ask!
